experiment/measurer.py

```diff
@@ -571,9 +571,8 @@ def set_up_coverage_binaries(pool, experiment):
     """Set up coverage binaries for all benchmarks in |experiment|."""
     # Use set comprehension to select distinct benchmarks.
     benchmarks = {
-        trial.benchmark for trial in db_utils.query(models.Trial).distinct(
-            models.Trial.benchmark).filter(
-                models.Trial.experiment == experiment)
+        trial.benchmark for trial in db_utils.query(models.Trial).filter(
+            models.Trial.experiment == experiment)
```

Comment:
I think the original query was incorrect and it should be:
```python
db_utils.query(models.Trial.benchmark).distinct()\
        .filter(models.Trial.experiment == experiment)
```
The original query returns trials, while we actually want the list of unique benchmarks. The above query should do that, so there's no need for the `trial.benchmark for trial in ...` part either.

You can see the raw SQL query by `print()`-ing the sqlachemy query (`db_utils.query`). The original query translates to:
```sql
SELECT DISTINCT ON (trial.benchmark) trial.id AS trial_id, trial.fuzzer AS trial_fuzzer, trial.experiment AS trial_experiment, trial.benchmark AS trial_benchmark, trial.time_started AS trial_time_started, trial.time_ended AS trial_time_ended, trial.preemptible AS trial_preemptible, trial.preempted AS trial_preempted 
FROM trial 
WHERE trial.experiment = %(experiment_1)s
```

While the one suggested above translates to what i believe we want here:

```sql
SELECT DISTINCT trial.benchmark AS trial_benchmark 
FROM trial 
WHERE trial.experiment = %(experiment_1)s
```

Reply:
Make sense and will test this suggestion.

Reply:
Here is the error log for this suggested change:
```
Process Process-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/work/src/experiment/measurer.py", line 79, in measure_loop
    set_up_coverage_binaries(pool, experiment)
  File "/work/src/experiment/measurer.py", line 579, in set_up_coverage_binaries
    pool.map(set_up_coverage_binary, benchmarks)
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 657, in get
    raise self._value
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 431, in _handle_tasks
    put(task)
  File "/usr/local/lib/python3.7/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/local/lib/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class 'sqlalchemy.orm.session.Session'>: it's not the same object as sqlalchemy.orm.session.Session
```

Looks like we can only query `models.Trial`, not `models.Trial.benchmark`.

Reply:
Can you post the change you made?
I suspect this error is unrelated to using `distinct` as laszlo suggested.
I'm guessing your query with `.list()` fixes this issue.

Reply:
Sure. 
```
     benchmarks = {
-        trial.benchmark for trial in db_utils.query(models.Trial).filter(
-            models.Trial.experiment == experiment)
+        db_utils.query(models.Trial.benchmark).filter(
+        models.Trial.experiment == experiment)
     }
```
Could you specify the suggested place for `.list()`?

Reply:
Heh, looks like I forget `.distinct()`. Let me add this and test again.

Reply:
I think you forgot `distinct`?
But anyway, i'm pretty sure now that the error you saw was from trying to put a query in a set, when really the results of the query should go in it.
Actually, we don't even need the set at all if we use `distinct`.
Try this:
`benchmarks = db_utils.query(models.Trial.benchmark).distinct().filter(models.Trial.experiment == experiment).list()`

Reply:
For 
```
benchmarks = db_utils.query(models.Trial.benchmark).distinct().filter(models.Trial.experiment == experiment).list()
```
It has
```
Process Process-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/work/src/experiment/measurer.py", line 79, in measure_loop
    set_up_coverage_binaries(pool, experiment)
  File "/work/src/experiment/measurer.py", line 574, in set_up_coverage_binaries
    models.Trial.experiment == experiment).list()
AttributeError: 'Query' object has no attribute 'list'
```

Reply:
For 
```
benchmarks = [ db_utils.query(models.Trial.benchmark).distinct().filter(models.Trial.experiment == experiment) ]
```
It has
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/work/src/experiment/measurer.py", line 79, in measure_loop
    set_up_coverage_binaries(pool, experiment)
  File "/work/src/experiment/measurer.py", line 578, in set_up_coverage_binaries
    pool.map(set_up_coverage_binary, benchmarks)
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 657, in get
    raise self._value
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 431, in _handle_tasks
    put(task)
  File "/usr/local/lib/python3.7/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/local/lib/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class 'sqlalchemy.orm.session.Session'>: it's not the same object as sqlalchemy.orm.session.Session
```

Reply:
`benchmarks = [ db_utils.query(models.Trial.benchmark).distinct().filter(models.Trial.experiment == experiment) ]` 
This doesn't work because that is not the right way to get a list of items in the query, this gives you a list containing just the query.
I think it is `.all()` instead of `.list()` but in any case, try: `benchmarks = list(db_utils.query(models.Trial.benchmark).distinct().filter(models.Trial.experiment == experiment))`.

Reply:
This
```
db_utils.query(models.Trial.benchmark).distinct().filter(models.Trial.experiment == experiment)
```
will return a list like this
```
[ ('string_content_1',), ('string_content_2',)]
```
Thus, the current commit will work. (033502c)